// @reclaimprotocol/reclaim-sdk/dist/ReclaimProtocol/Template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../config");
const utils_1 = require("../utils");
/** Template instance */
class TemplateInstance {
    /**
     * Constructor
     * @param {Template} template
    */
    constructor(template, regexes) {
        this._template = template;
        this._regexes = (0, utils_1.encodeBase64)(regexes);
    }
    /**
     * Getter template
     * @return {Template}
    */
    get template() {
        return this._template;
    }
    /**
     * Getter id
     * @return {string}
     */
    get id() {
        return this._template.id;
    }
    /**
     * Getter callbackId
     * @return {string}
     */
    get callbackId() {
        return (0, utils_1.getCallbackIdFromUrl)(this._template.callbackUrl);
    }
    /**
     * Getter template url
     * @return {string}
    */
    get reclaimUrl() {
        return config_1.RECLAIM_APP_URL + encodeURIComponent(JSON.stringify(this._template));
    }
    /**
     * Getter regexes: base64 encoded regexes
     * @return {string}
     */
    get expectedProofsInCallback() {
        return this._regexes;
    }
}
exports.default = TemplateInstance;
